Reformat inline-assembly chapter as a spec chapter. (#1523)
rust-lang/reference: src/inline-assembly.md

# Inline assembly [asm]

r[asm.macros]
The macros [`core::arch::asm!`] and [`core::arch::global_asm!`] expand to inline assembly syntax when used in the expression position and item position respectively. The macros shall not be expanded in any other context.

> [!NOTE]
> The expansion of the macros has no stable syntax equivalent. This section will refer to the expansion of the macro, rather than the surface syntax.

r[asm.safety]

> The macro [`core::arch::asm!`] shall be expanded only within an `unsafe` block.

Author: Does it? Shall is a static constraint on the program, so program is ill-formed if violated.

> The macro [`core::arch::asm!`] shall be expanded only within an `unsafe` block.

Reviewer: As a casual reader, I agree that it could stand to be rephrased to mention a compiler error, or state some similarity to calling an `unsafe fn`.

Though the below compile_fail test, which you've asked to have stylistically called out as "this will fail to compile", might serve to alleviate the concern a little bit here. But it would be nice if this sentence could avoid being misleading in the first place.

sym_operand := [ident "="] "sym" sym_expr
clobber_abi := "clobber_abi(" <abi> *("," <abi>) [","] ")"
option := "pure" / "nomem" / "readonly" / "preserves_flags" / "noreturn" / "nostack" / "att_syntax" / "raw"
options := "options(" option *("," option) [","] ")"
operand := reg_operand / sym_operand / clobber_abi / options
asm_inner := format_string *("," format_string) *("," operand) [","]
asm := "asm!(" asm_inner ")"
global_asm := "global_asm!(" asm_inner ")"

non_format_char := ANY_CHAR // except "{" and "}"
operand_specifier := ident / DEC_LITERAL
expansion_specifier := *non_format_char
format_specifier := "{" [operand_specifier] [":" *expansion_specifier]  "}"
format_escape := "{{" / "}}"
asm_string_piece := non_format_char / format_specifier / format_escape
asm_string_content := [*asm_string_piece]
```

## Invocation [asm.invocation]

r[asm.invocation.asm]
The [`core::arch::asm!`] macro shall be expanded in an expression context only. The input tokens shall match the `asm_inner` production. The expansion is [`unsafe`][static.expr.safety] and has type `()`, unless the option `noreturn` is specified, in which case it has type `!`.

```rust

> # #[cfg(target_arch = "x86_64")] 

Reviewer 2: This method of using `cfg` along with an `fn main` confuses rustdoc and causes this to fail on non-x86_64 architectures. Can you look into some way to work around that issue?

For this particular example, does it need the cfg?

Author: It does at least for the 5 architectures that asm is valid on (since if it gets tested off of those architectures, it will just error out)

> pub fn main(){

Reviewer 2: ```suggestion
pub fn main() {
```

It would be good to generally try to ensure that the examples are formatted in a way that is pretty close to how rustfmt works.

r[asm.invocation.concat]
If multiple `format_string` inputs are provided, then they are concatenated as though by the [`core::concat!`] macro, separating each `format_string` with a string containing a single newline character. If any `format_string` begins a `format_specifier` that is not terminated before the end of the `format_string`, the program is ill-formed. The resulting string is known as the *joined asm-string*

```rust
# #[cfg(target_arch = "x86_64")] {unsafe{
let mut x: i32;
// The following lines are equivalent
core::arch::asm!("mov rax, 5", "mov rcx, rax", out("rax") x, out("rcx") _); 
core::arch::asm!("mov rax, 5\nmov rcx, rax", out("rax") x, out("rcx") _);  
# }}
```

r[asm.invocation.operands]
Each operand, other than an explicit register operand ([asm.operands.register]) shall be mentioned by at least one format_specifier in the *joined asm-string*. Explicit registers may not be referred to be a format_specifier.

> ```rust,compile_fail

Reviewer 2: These compile_fail annotations won't work with non-x86_64 architectures (since it won't fail to compile).

Author: Yeah, I'll add a
```
# #[cfg(not(target_arch = "x86_64")]
# core::compile_error!("Tests currently require x86_64. Please add per-architecture tests to run off of x86_64");

# }
```

> ## Syntax [asm.syntax]

Reviewer 2: It's a little unclear to me what these bracketed things are in the headers. Can you say more about why you added those, or what their intent is?

Like here, I would expect this to be something like:

```
## Syntax

r[asm.syntax]
```

Author: That seems to me like it would be formatted weird. The Rule would appear dangling, not attached to anything.

The rules in the header are also supposed to be the anchor for the entire header section (although I think I got the cmark syntax wrong for that).

Reviewer 3: What does the `r` in front of the bracket refer to? Rule? Render? Where in the md-book documentation can I find that - I tried to search, but couldn't find it?

I am pretty indifferent on which way we go here. The anchor will still basically go to the top even if we format it as @ehuss suggests. I could almost argue another way that it should *above* the header similar to how we do other things.

```
r[asm.syntax]
## Syntax
```

But I don't feel strongly. Let's just pick something and go with it for now.

Author: Rule I believe - it's in mdbook-spec not mdbook proper.

Putting the anchor above doesn't quite do what I want, though - I want to replace the default anchor for the header, and I also don't really want the rendering of the rule in this case - it seems like it's supposed to have normal text below it (not a header or a code block), but we either hard switch the context with a header or we put another rule in immediately for the  next actual text entry. That's what I meant by "The Rule would appear dangling".

Reviewer 2: > Where in the md-book documentation can I find that - I tried to search, but couldn't find it?

BTW, it is documented here: https://github.com/rust-lang/reference/blob/master/docs/authoring.md#rules

Author: I've changed this to have the rule after. I'd like to have a more extended discussion about this, but it's a mechanical change to revert it anyways so I'm not going to block this PR on it.  I'll open a policy issue for it though, and would like to put it on the agenda for tomorrow. Should that issue go on rust-lang/reference or rust-lang/spec?

Author: File [spec#71](https://github.com/rust-lang/spec/issues/71)

```rust
# #[cfg(target_arch = "x86_64")] { unsafe{
let x: i32;
core::arch::asm!("mov {output}, {input}", input = in(reg) 5i64, output = out(reg) x);
# }}
```

r[asm.invocation.expansion]
The *joined asm-string* is expanded as defined in [asm.operands.expansion], replacing each `format_specifier` with the appropriate expansion for the operand. The resulting string is called the *expanded asm-string*.

r[asm.invocation.syntax]
The syntax of the *expanded asm-string* is a subset of the GNU AS syntax for the target. Invoking the macro with a *expanded asm-string* that does not match syntax requirements is *conditionally supported* and has *assembler dependent behaviour*. Invoking a directive that is not specified by [asm.directives] is *conditionally supported* and has *assembler dependent behaviour*.

> [!TARGET-SPECIFIC]
> On x86 and x86_64 targets, the syntax of the *expanded asm-string* acts as though the directive `.intel_syntax noprefix` is issued before parsing the *expanded asm-string*, except that the `option(att_syntax)` causes the syntax to act as though the directive `.att_syntax prefix` is issued before parsing the *expanded asm-string* instead.

> > On ARM and Aarch64 targets, the syntax of the *expanded asm-string* acts as though the directive `.syntax unified` is issued before parsing the *expanded asm-string*.

Reviewer 4: That's only on ARM, not AArch64.

r[asm.invocation.duplication]
The number of times, locations, and the order in which a given invocation of [`core::arch::asm!`] is expanded is unspecified.

```rust,ignore
// The following code may have suprising results, and may fail to compile or link. 
// The results, including whether it succesfully compiles, may depend on non-local use sites of the function, and on optimization settings.
# #[cfg(target_arch = "x86_64")] { unsafe{
let x: i32;
core::arch::asm!("foo: jmp foo");
# }}
```

> [!NOTE]
> In particular, an asm block may be duplicated, for example if the containing function is inlined, or omitted from the output entirely.
> As a consequence, asm blocks should not use directives that have non-idempotent non-local effects, or named labels and symbol definitions. 
> Additionally, two asm blocks may not rely upon being adjacent in executable memory, even if they are adjacent in the source.

> [!NOTE]

> > Local Labels (a decimal literal that doesn't solely consist of 0s and 1s) may be used freely if the asm block needs to define a label.

Author: I think it should be noted so that people don't rely on the statement to say "But it says 'a decimal literal'", but it could be mentioned as being a bug. We can always loosen the specification, but we can't tighten it later after everything is all stabilized.

r[asm.invocation.directive-state]
The *expanded asm-string* shall not issue a directive that modifies the global state of the assembler for processing inputs unless it issues a directive to restore that state it had upon entering the block. No diagnostic is required.

> [!NOTE]
> This include state such as the current section of the assembler, the syntax mode, or the kind of assembly output being generated.
> Failing to obey this requirement can have significant impact on code generation, including code unrelated to the asm block. For example, an asm block that issues a `.data` directive without resetting to the appropriate section for the function can cause the following code in the function to be generated in the `.data` section, and for execution to fall off the asm block into improper memory.

r[asm.invocation.global-section]
The *expanded asm-string* of a [`core::arch::global_asm!`] invocation acts as though a `.section` directive is issued before the *expanded asm-string*  which causes code to be generated in the default section on the target for executable code.

> [!NOTE]
> This section is typically named `.text`. 

r[asm.invocation.prefix-instr]

> An *expanded asm-string* shall not end with an instruction that is interpreted as a prefix on the architecture. No Diagnostic is required

Author: I didn't see a complete resolution in that thread (@joshtriplett seemed skeptical that it should be considered DB to have it in dead code), so I'm leaving it as the most restrictive rule. We can change the guarantees here if we find it handled correctly (or fixable), but this reformatting isn't intended to change the rules, just change how they are expressed.

> An *expanded asm-string* shall not end with an instruction that is interpreted as a prefix on the architecture. No Diagnostic is required

Reviewer 4: That's fair, we can update it in a later PR.

```rust
# #[cfg(target_arch = "x86_64")] { unsafe{
let mut x: i32;
core::arch::asm!("mov rax, {}", in(reg) 5i64, out("eax") x);
core::arch::asm!("mov {1}, {0}", in(reg) 5i64, out(reg) x);
# }}
```

r[asm.operands.named]
Operands that specify an ident are named operands. A named operand shall not specify an explicit register `reg_spec`. Named operand specifiers may be referred to only by named operand specifiers.
```rust
# #[cfg(target_arch = "x86_64")] { unsafe{
let x: i32;
core::arch::asm!("mov rax, {input}", input = in(reg) 5i64, out("eax") x);
# }}
```

r[asm.operands.registers]
Operands that specify an explicit register `reg_spec` are explicit register operands. 

```rust
# #[cfg(target_arch = "x86_64")] { unsafe{
let x: i32;
core::arch::asm!("mov eax, ecx", in("rcx") 5i64, out("eax") x);
# }}
```

> [!NOTE]
> Explicit Register Operands have no `ident` name and cannot be referred to by an operand specifier

r[asm.operands.types]

> Each operand, other than a placeholder expression shall be of an integer type, floating-point type, function pointer type, pointer type, or target-specific vector type. These types are collectively called *asm operand types*. A pointer type is an *asm operand type* only if the pointee type has no metadata-type.

Author: I can add that `input_expr`s have to be coerced to an *asm operand type*.

> Each operand, other than a placeholder expression shall be of an integer type, floating-point type, function pointer type, pointer type, or target-specific vector type. These types are collectively called *asm operand types*. A pointer type is an *asm operand type* only if the pointee type has no metadata-type.

Author: (Aside, is that a way I can force a coercion to a function pointer without knowing the signature :eyes:)

> Each operand, other than a placeholder expression shall be of an integer type, floating-point type, function pointer type, pointer type, or target-specific vector type. These types are collectively called *asm operand types*. A pointer type is an *asm operand type* only if the pointee type has no metadata-type.

Reviewer 4: At the moment we specifically only perform the 2 coercions I mentioned: https://github.com/rust-lang/rust/blob/f00f85091904530ee8f5b29251c53571e0062a76/compiler/rustc_hir_typeck/src/expr.rs#L3253-L3272

Also we accept `MaybeUninit` versions of all allowed types: https://github.com/rust-lang/rust/pull/114790

> [!NOTE]
> The name given to an item used by a sym-expr that does not have a linkage name may be known as the "mangled" name of the item.

> [!NOTE]
> A sym operand does not include any relocation modifiers such as `@plt` or `@tpoff`. The *joined asm-string* is responsible for including these as required.

> [!TARGET-SPECIFIC]
> On x86 and x86_64 targets, the register name is expanded as-is if the `options(att_syntax)` is not used, and with the `%` prefix if `options(att_syntax)` is used. 

r[asm.operands.global]
The program shall not use an operand, other than a sym operand, in the expansion of the [`core::arch::global_asm!`] macro.

<!--TODO: Test `global_asm!`-->

```rust,compile_fail,ignore
# #[cfg(target_arch = "x86_64")]
core::arch::global_asm!("", in("eax") 5);
# #[cfg(not(target_arch = "x86_64"))] compile_error!("Inline Assembly Tests are not supported off of x86_64");
```

```rust,ignore
static FOO: () = ();
# #[cfg(target_arch = "x86_64")]
core::arch::global_asm!("/*{}*/", sym FOO);
```

r[asm.operands.clobbers_abi]

> A special operand `clobbers_abi` may be specified. If the `clobers_abi` operand is specified, the no reg_operand, other than an *explicit register operand*, shall be specified. When specified, it accepts a string literal which shall belong to a subset of the string literals accepted for an `extern` calling convention specification. The `clobbers_abi` special operand acts as though it is replaced by a `lateout` operand with an out-expr of `_` for each register considered by the specified calling convention to not be preserved by a function call. 

Author: I mention that it belongs to a subset, not any valid extern string already.

> shall belong to a subset of the string literals accepted for an `extern` calling convention specification

> | ARM | `dreg_low8` | `d[0-8]` | `vfp2` | `i64`, `f64`, `i8x8`, `i16x4`, `i32x2`, `i64x1`, `f32x2` |
> | ARM | `qreg` | `q[0-15]` | `neon` | `i8x16`, `i16x8`, `i32x4`, `i64x2`, `f32x4` |
> | ARM | `qreg_low8` | `q[0-7]` | `neon` | `i8x16`, `i16x8`, `i32x4`, `i64x2`, `f32x4` |
> | ARM | `qreg_low4` | `q[0-3]` | `neon` | `i8x16`, `i16x8`, `i32x4`, `i64x2`, `f32x4` |
> | RISC-V32 | `reg` | `x1`, `x[5-7]`, `x[9-15]`, `x[16-31]` (non-RV32E) | None | `i8`, `i16`, `i32`, `f32` |
> | RISC-V64 | `reg` | `x1`, `x[5-7]`, `x[9-15]`, `x[16-31]`  | None | `i8`, `i16`, `i32`, `f32`, `i64`, `f64` |
> | RISC-V | `freg` | `f[0-31]` | `f` | `f32`, `f64` (requires `d`) |
> | RISC-V | `vreg` | `v[0-31]` | N/A | Only clobbers |
> | LoongArch | `reg` | `$r1`, `$r[4-20]`, `$r[23,30]` | None | `i8`, `i16`, `i32`, `i64`, `f32`, `f64` |
> | LoongArch | `freg` | `$f[0-31]` | None | `f32`, `f64` |

> **Notes**:
> - On x86 we treat `reg_byte` differently from `reg` because the compiler can allocate `al` and `ah` separately whereas `reg` reserves the whole register.
>
> - On x86-64 the high byte registers (e.g. `ah`) are not available in the `reg_byte` register class.
>

r[asm.register.clobbers_only]

> Certain registers and register classes are *clobbers only*. Such register names or register classes shall not be specified by an operand, other than a clobbers output.

Author: It's the register class that is specified as "Clobbers Only". Although I probably should restrict clobbers output in the operands section to "Shall be an *explicit register operand*"

* A pointer tag which has been exposed, 
* A pointer tag which was passed as an input operand, or
* A pointer tag which is accessible by reading any memory the asm block can read under this clause, recursively.

r[asm.evaluation.unwind]
The behaviour is undefined if an inline assembly block exits by unwinding from a panic or a foreign exception.

```rust,ignore
// The following snippet has undefined behaviour
extern "C-unwind" fn panics(){panic!("unwind through asm")}
# #[cfg(target_arch = "x86_64")] { unsafe{
core::arch::asm!("call {}", sym panics);
# }}
```

r[asm.evaluation.register-value]
The behaviour is undefined upon exiting an asm block unless the stack pointer register and each operand-usable register not mentioned by an `out` , `lateout`, `inout`, or `inlateout` operand has the value the register held upon entry to the asm block.

> [!TARGET-SPECIFIC]
> In addition to operand-usable registers, certain other registers on a target may require being preserved, or have specific rules regarding the value at exit.
> On x86 and x86-64 targets:
> * The Direction flag (`flags.DF`) is clear upon entry and must be clear upon exit

> > * The x87 Stack (that is the `TOP` field of the floating-point status word, and each bit in the floating-point tag word) must be preserved and restored upon exit. If all x87 `st` registers are marked as clobbered, the stack is guaranteed to be empty on entry to the asm block (that is, `TOP` is set to `0x7` and the `ftw` is set to `0xFFFF`).

Author: This is specifically for architectural register state. Other state is/would be handled by other rules.

```rust
# #[cfg(target_arch = "x86_64")] { unsafe{
let x: i32;
core::arch::asm!("mov {:e}, %eax", in(reg) 5, out("eax") x, options(att_syntax));
# }}
```

r[asm.options.nomem]
The `nomem` option may be specified. The behaviour is undefined if the assembly block modifies any allocation, disables, freezes, or activates any tag, *synchronizes-with* any other thread of execution or signal handler, and the implementation may assume that the behaviour or outputs of the assembly block does not depend on the contents of any allocation.

```rust,ignore
// The following snippet has undefined behaviour
static mut FOO: i32 = 5;
# #[cfg(target_arch = "x86_64")] { unsafe{
core::arch::asm!("mov dword ptr [{}+rip], 3", sym FOO, options(nomem));
# }}
```

```rust
// The following snippet may have unpredictable results
static mut FOO: i32 = 5;
# #[cfg(target_arch = "x86_64")] { unsafe{
let x: i32;
core::arch::asm!("mov {:e}, dword ptr [{}+rip]", out(reg) x, sym FOO, options(nomem));
# }}
```

r[asm.options.readonly]

> The `readonly` option may be specified. The behaviour is undefined if the assembly block modifies any allocation or activates any tag. 

Reviewer 4: `readonly` also isn't allowed to synchronize with anything. For example you can't use `readonly` for a fence instruction.

Author: Are `acquire` operations not allowed for `readonly`? That seems odd to me.

Reviewer 4: I'm not 100% sure if it's legal for the compile to eliminate an acquire load if its result is unused. In fact LLVM currently doesn't do this even for relaxed loads.

Author: It is, because it can just take the value from the last store that already *happens-before* so it doesn't make any new side effects visible. LLVM is just understably conservative arround optimizing atomics.

```rust,ignore
# #[cfg(target_arch = "x86_64")] { unsafe{
core::arch::asm!("cmp eax, eax", in("eax") 5, options(preserve_flags));
# }}
```

r[asm.options.noreturn]
The `noreturn` option may be specifed. An invocation of the [`core::arch::asm!`] macro that specifies the `noreturn` option expands to an expression of type `!`. The behaviour is undefined if an evaluation of the assembly block exits. The program shall not specify the `clobber_abi` specification, or an operand that is an `out`, `lateout`, `inout`, or `inlateout` operand. 

```rust
# #[cfg(target_arch = "x86_64")]
pub fn main() -> ! {
  unsafe{
    core::arch::asm!("xor edi, edi", "call exit@plt", options(noreturn));
  }
}
```

```rust,ignore
// The following snippet has undefined behaviour
# #[cfg(target_arch = "x86_64")] { unsafe{
core::arch::asm!("", options(noreturn));
# }}
```

```rust,compile_fail
# #[cfg(target_arch = "x86_64")] { unsafe{
let x: i32;
core::arch::asm!("xor edi, edi", "call exit@plt", out("edi") x, options(noreturn));
# }}
# #[cfg(not(target_arch = "x86_64"))] compile_error!("Inline Assembly Tests are not supported off of x86_64");
```

r[asm.options.global]

> A program shall not specify an option, other than the `att_syntax` option, in an invocation of the [`core::arch::global_asm!`] macro.

Reviewer 4: You can also use the `raw` option with global_asm. In fact it seems the `raw` option was omitted entirely in the spec?

> [!NOTE]
> The name given to an item used by a sym-expr that does not have a linkage name may be known as the "mangled" name of the item.

> [!NOTE]
> A sym operand does not include any relocation modifiers such as `@plt` or `@tpoff`. The *joined asm-string* is responsible for including these as required.

> [!TARGET-SPECIFIC]
> On x86 and x86_64 targets, the register name is expanded as-is if the `options(att_syntax)` is not used, and with the `%` prefix if `options(att_syntax)` is used. 

r[asm.operands.global]
The program shall not use an operand, other than a sym operand, in the expansion of the [`core::arch::global_asm!`] macro.

<!--TODO: Test `global_asm!`-->

```rust,compile_fail,ignore
# #[cfg(target_arch = "x86_64")]
core::arch::global_asm!("", in("eax") 5);
# #[cfg(not(target_arch = "x86_64"))] compile_error!("Inline Assembly Tests are not supported off of x86_64");
```

```rust,ignore
static FOO: () = ();
# #[cfg(target_arch = "x86_64")]
core::arch::global_asm!("/*{}*/", sym FOO);
```

r[asm.operands.clobbers_abi]

> A special operand `clobbers_abi` may be specified. If the `clobers_abi` operand is specified, the no `out`, `lateout`, `inout`, or `inlateout` reg_operand, other than an *explicit register operand*, shall be specified. When specified, it accepts a string literal which shall belong to a subset of the string literals accepted for an `extern` calling convention specification. The `clobbers_abi` special operand acts as though it is replaced by a `lateout` operand with an out-expr of `_` for each register considered by the specified calling convention to not be preserved by a function call. 

Reviewer 4: ```suggestion
A special operand `clobbers_abi` may be specified. If the `clobers_abi` operand is specified, then no `out`, `lateout`, `inout`, or `inlateout` reg_operand, other than an *explicit register operand*, shall be specified. When specified, it accepts a string literal which shall belong to a subset of the string literals accepted for an `extern` calling convention specification. The `clobbers_abi` special operand acts as though it is replaced by a `lateout` operand with an out-expr of `_` for each register considered by the specified calling convention to not be preserved by a function call. 
```

reg_spec := <register class> / "\"" <explicit register> "\""
input_expr := expr
output_expr := expr / "_"
inout_expr := input_expr ["=>" output_expr]
operand_expr := input_expr / output_expr / inout_expr
sym_expr := path
reg_operand := [ident "="] dir_spec "(" reg_spec ")" operand_expr
sym_operand := [ident "="] "sym" sym_expr
clobber_abi := "clobber_abi(" <abi> *("," <abi>) [","] ")"
option := "pure" / "nomem" / "readonly" / "preserves_flags" / "noreturn" / "nostack" / "att_syntax" / "raw"
options := "options(" option *("," option) [","] ")"
operand := reg_operand / sym_operand / clobber_abi / options
asm_inner := format_string *("," format_string) *("," operand) [","]
asm := "asm!(" asm_inner ")"
global_asm := "global_asm!(" asm_inner ")"

non_format_char := ANY_CHAR // except "{" and "}"
operand_specifier := ident / DEC_LITERAL
expansion_specifier := *non_format_char
format_specifier := "{" [operand_specifier] [":" *expansion_specifier]  "}"
format_escape := "{{" / "}}"
asm_string_piece := non_format_char / format_specifier / format_escape
asm_string_content := [*asm_string_piece]
```

## Invocation 

r[asm.invocation]

r[asm.invocation.asm]

> The [`core::arch::asm!`] macro shall be expanded in an expression context only. The input tokens shall match the `asm_inner` production. The expansion is [`unsafe`][static.expr.safety] and has type `()`, unless the option `noreturn` is specified, in which case it has type `!`.

Reviewer 2: This link to `unsafe` is broken since there aren't any rules for that, yet. Can you make sure to link directly until those sections are ready? In this case,

```
[`unsafe`]: unsafety.md
```

# }}
```

r[asm.invocation.explicit-positional]
A `format_specifier` that has an `operand_specifier` which is a DEC_LITERAL is called an explicit positional specifier, and refers to the `nth` successive positional operand, where `n` is the value of the DEC_LITERAL.

```rust
# #[cfg(target_arch = "x86_64")] { unsafe{
let x: i32;
core::arch::asm!("mov {1}, {0}", in(reg) 5i64, out(reg) x);
# }}
```

r[asm.invocation.named]
A `format_specifier` that has an `operand_specifier` which is an ident is called a named specifier, and refers to the named operand with the specified name.

```rust
# #[cfg(target_arch = "x86_64")] { unsafe{
let x: i32;
core::arch::asm!("mov {output}, {input}", input = in(reg) 5i64, output = out(reg) x);
# }}
```

r[asm.invocation.expansion]
If the `raw` option is not specified, the *joined asm-string* is expanded as defined in [asm.operands.expansion], replacing each `format_specifier` with the appropriate expansion for the operand. The resulting string is called the *expanded asm-string*. If the `raw` option is specified, the *expanded asm-string* is the *joined asm-string* verbatim. 

r[asm.invocation.syntax]

> The syntax of the *expanded asm-string* is a subset of the GNU AS syntax for the target. Invoking the macro with a *expanded asm-string* that does not match syntax requirements is *conditionally supported* and has *assembler dependent behaviour*. Invoking a directive that is not specified by [asm.directives] is *conditionally supported* and has *assembler dependent behaviour*.

Reviewer 2: Can you make sure to use American English? (Here and elsewhere in the document.)

```suggestion
The syntax of the *expanded asm-string* is a subset of the GNU AS syntax for the target. Invoking the macro with a *expanded asm-string* that does not match syntax requirements is *conditionally supported* and has *assembler dependent behavior*. Invoking a directive that is not specified by [asm.directives] is *conditionally supported* and has *assembler dependent behavior*.
```

> [!TARGET-SPECIFIC]
> On x86 and x86_64 targets, the syntax of the *expanded asm-string* acts as though the directive `.intel_syntax noprefix` is issued before parsing the *expanded asm-string*, except that the `option(att_syntax)` causes the syntax to act as though the directive `.att_syntax prefix` is issued before parsing the *expanded asm-string* instead.
> On ARM targets, the syntax of the *expanded asm-string* acts as though the directive `.syntax unified` is issued before parsing the *expanded asm-string*.

r[asm.invocation.duplication]
The number of times, locations, and the order in which a given invocation of [`core::arch::asm!`] is expanded is unspecified.

```rust,ignore

> // The following code may have suprising results, and may fail to compile or link. 
> // The results, including whether it succesfully compiles, may depend on non-local use sites of the function, and on optimization settings.

Reviewer 2: Can you make sure that the code examples are wrapped roughly at 80 columns? The way it gets presented in the book requires horizontal scrolling which can be awkward.

Also, there are some misspellings here.

```suggestion
// The following code may have surprising results, and may fail to compile or
// link. The results, including whether it successfully compiles, may depend on
// non-local use sites of the function, and on optimization settings.
```

r[asm.invocation.directive-state]
The *expanded asm-string* shall not issue a directive that modifies the global state of the assembler for processing inputs unless it issues a directive to restore that state it had upon entering the block. No diagnostic is required.

> [!NOTE]
> This include state such as the current section of the assembler, the syntax mode, or the kind of assembly output being generated.
> Failing to obey this requirement can have significant impact on code generation, including code unrelated to the asm block. For example, an asm block that issues a `.data` directive without resetting to the appropriate section for the function can cause the following code in the function to be generated in the `.data` section, and for execution to fall off the asm block into improper memory.

r[asm.invocation.global-section]
The *expanded asm-string* of a [`core::arch::global_asm!`] invocation acts as though a `.section` directive is issued before the *expanded asm-string*  which causes code to be generated in the default section on the target for executable code.

> [!NOTE]
> This section is typically named `.text`. 

r[asm.invocation.prefix-instr]

> An *expanded asm-string* shall not end with an instruction that is interpreted as a prefix on the architecture. No Diagnostic is required

Reviewer 2: ```suggestion
An *expanded asm-string* shall not end with an instruction that is interpreted as a prefix on the architecture. No diagnostic is required.
```

Also, I'm not sure why it is mentioning whether a diagnostic is required. Can you say more why it says that?

Author: The presumption is that a diagnostic is required, and that the compiler is forbidden to compile an ill-formed program (one that violates a constraint like that one). That overrides the rule by telling the compiler that it's free to ignore the violation and translate the program notwithstanding, but that the result of the program is allowed to be anything.

[`__m256`]: https://doc.rust-lang.org/core/arch/x86_64/struct.__m256.html

```rust
# #[cfg(target_arch = "x86_64")] { unsafe{
# use core::arch::x86_64::__m128;
// But vector types are allowed.
let x: __m128;
core::arch::asm!("xorps xmm0, xmm0", out("xmm0") x);
# }}
```

r[asm.operands.input-coerceable-types]
Each reference type, where the pointee type has no metadata-type, and each function item type are collectively called *input coerceable types*.

```rust
# #[cfg(target_arch = "x86_64")] { unsafe{
let x = 5;
let y: i32;
core::arch::asm!("mov eax, dword ptr [{}]", in(reg) &x, out("eax") y); // equivalent to asm!("mov eax, dword ptr [{}]", in(reg) (&x) as *const i32, out("eax") y); 
#}}
```

```rust,compile_fail
# #[cfg(target_arch = "x86_64")] { unsafe{
let y: &mut i32;
core::arch::asm!("mov {}, 0", out(reg) 5); 
#}}
# #[cfg(not(target_arch = "x86_64"))] compile_error!("Inline Assembly Tests are not supported off of x86_64");
```

r[asm.operands.in-expr]
An `input_expr` shall be a value expression of an *asm operand type* or an *input coerceable type*. If the expression is of an *input coerceable type*, it is coerced to an *asm operand type*. 

r[asm.operands.out-expr]

> An `output_expr` shall be the placeholder expression `_` or a (potentially unitialized) place expression of an *asm operand type*. If the place expression is initialized, it shall be a mutable place.

Reviewer 2: ```suggestion
An `output_expr` shall be the placeholder expression `_` or a (potentially uninitialized) place expression of an *asm operand type*. If the place expression is initialized, it shall be a mutable place.
```

```rust
# #[cfg(target_arch = "x86_64")] { unsafe{
let x: i32 = 5;
let y: i32;
core::arch::asm!("mov {:e}, {:e}", out(reg) y, in(reg) x);
# }}
```

> [!NOTE]
> The supported modifiers are a subset of LLVM's (and GCC's) [asm template argument modifiers][llvm-argmod], but do not use the same letter codes.

> [!NOTE]
> - on ARM `e` / `f`: this prints the low or high doubleword register name of a NEON quad (128-bit) register.
> - on x86: our behavior for `reg` with no modifiers differs from what GCC does.
>   GCC will infer the modifier based on the operand value type, while we default to the full register size.
> - on x86 `xmm_reg`: the `x`, `t` and `g` LLVM modifiers are not yet implemented in LLVM (they are supported by GCC only), but this should be a simple change.

r[asm.template.diagnostic]
A lint diagnostic should be emitted if a modifier is omitted, or a modifier is used such that the modified expanded register is of an inappropriate width for the type used to initialize the operand

[llvm-argmod]: http://llvm.org/docs/LangRef.html#asm-template-argument-modifiers

## Behaviour of an asm block 

r[asm.evaluation]

r[asm.evaluation.general]

> Each evaluation of an asm block (invocation of [`core::arch::asm!`]) shall perform an operation that correpsonds to the result of a valid sequence of operations on the Minirust Abstract Machine. The behaviour is undefined if the operations performed by the asm block do not validly correspond to a valid sequence of Minirust operations.

Reviewer 2: ```suggestion
Each evaluation of an asm block (invocation of [`core::arch::asm!`]) shall perform an operation that corresponds to the result of a valid sequence of operations on the Minirust Abstract Machine. The behavior is undefined if the operations performed by the asm block do not validly correspond to a valid sequence of Minirust operations.
```

> [!NOTE]
> The operation the asm block performs may differ between evaluations of the same asm block.

> [!TARGET-SPECIFIC]

> > The correspondance between the operation performed by the asm block is target-dependant and implementation-dependant, subject to the rules set in [asm.operands]. Unless the program modifies the execution state, the basic operation performed by the asm block is the one performed by executing the sequence of instructions specified in the *expanded asm-string* starting with the first instruction.

Reviewer 2: ```suggestion
> The correspondence between the operation performed by the asm block is target-dependant and implementation-dependant, subject to the rules set in [asm.operands]. Unless the program modifies the execution state, the basic operation performed by the asm block is the one performed by executing the sequence of instructions specified in the *expanded asm-string* starting with the first instruction.
```

r[asm.evaluation.reg-values]
The value of each register mentioned in an input operand is set according to [asm.operands] before evaluating any instructions in the asm block. The value of each other *operand-usable register* is unspecified. The value of all other registers is target-dependant.

> [!NOTE]
> The target may define that the register value (or some portion thereof) is undefined.

r[asm.evaluation.constraints]

> Certain constraints may be placed on the asm block, and on the requirements of the correspondance, by default or by an option explicitly specified on the asm block. The behaviour is undefined if any such constraint is violated.

Reviewer 2: ```suggestion
Certain constraints may be placed on the asm block, and on the requirements of the correspondence, by default or by an option explicitly specified on the asm block. The behaviour is undefined if any such constraint is violated.
```

Also, in general, I don't understand what "requirements of the correspondance" means.

Also, I'm in general not understanding this rule. What constraints is it referring to? Can you point to the original chapter that this is referring to?

Author: > Can you point to the original chapter that this is referring to?

This is the formalization of the "Rule of Inline Assembly" section's preamble. The constraints mentioned there are then elaborated in the rest of the section (for default constraints) and in [asm.options] (for option specific constraints). 

The correspondence is the correspondence between the abstract machine state and the physical machine state affected by the inline assembly - defined in the previous clause and target-specific admonition, and constrained in various sections (such as in [asm.operands] and [asm.registers]). IE. when you do X in assembly, what effect does that have on the Rust/Minirust AM, and how do we overall map both incoming state and outgoing state. The constraints then limit the results of the correspondence. For example, if the correspondence says that the operation wrote to some allocation, but the `nomem` option was set, you've violated that constraint on the asm block. Another example is the fact that if a register isn't mentioned by an `out`, `lateout`, `inout`, or `inlateout` operand, it must be preserved by the asm block. An asm block that then clobbers a register that is required to be preserved violates that constraint.

# #[cfg(target_arch = "x86_64")] { unsafe{
core::arch::asm!("call {}", sym panics);
# }}
```

r[asm.evaluation.register-value]
The behaviour is undefined upon exiting an asm block unless the stack pointer register and each operand-usable register not mentioned by an `out` , `lateout`, `inout`, or `inlateout` operand has the value the register held upon entry to the asm block.

> [!TARGET-SPECIFIC]
> In addition to operand-usable registers, certain other registers on a target may require being preserved, or have specific rules regarding the value at exit.
> On x86 and x86-64 targets:
> * The Direction flag (`flags.DF`) is clear upon entry and must be clear upon exit
> * The x87 Stack (that is the `TOP` field of the floating-point status word, and each bit in the floating-point tag word) must be preserved and restored upon exit. If all x87 `st` registers are marked as clobbered, the stack is guaranteed to be empty on entry to the asm block (that is, `TOP` is set to `0x7` and the `ftw` is set to `0xFFFF`).

## Options 

r[asm.options]

r[asm.options.general]
An options-spec provided in the asm invocation places constraints on the assembly block. 

r[asm.options.att_syntax]

> The `att_syntax` option may be specfied on the x86 and x86_64 target. The program shall not specify the `att_syntax` option on any other target.

Reviewer 2: ```suggestion
The `att_syntax` option may be specified on the x86 and x86_64 target. The program shall not specify the `att_syntax` option on any other target.
```

// The following snippet has undefined behaviour
static mut FOO: i32 = 5;
# #[cfg(target_arch = "x86_64")] { unsafe{
core::arch::asm!("mov dword ptr [{}+rip], 3", sym FOO, options(nomem));
# }}
```

```rust
// The following snippet may have unpredictable results
static mut FOO: i32 = 5;
# #[cfg(target_arch = "x86_64")] { unsafe{
let x: i32;
core::arch::asm!("mov {:e}, dword ptr [{}+rip]", out(reg) x, sym FOO, options(nomem));
# }}
```

r[asm.options.readonly]
The `readonly` option may be specified. The behaviour is undefined if the assembly block modifies any allocation or activates any tag. 

```rust,ignore
// The following snippet has undefined behaviour
static mut FOO: i32 = 5;
# #[cfg(target_arch = "x86_64")] { unsafe{
core::arch::asm!("mov dword ptr [{}+rip], 3", sym FOO, options(readonly));
# }}
```

r[asm.options.exclusive]
The program shall not specify both the `nomem` and `readonly` options.

```rust,compile_fail
# #[cfg(target_arch = "x86_64")] { unsafe{
core::arch::asm!("mov dword ptr [FOO+rip], 3", options(readonly, nomem));
# }}
# #[cfg(not(target_arch = "x86_64"))] compile_error!("Inline Assembly Tests are not supported off of x86_64");
```

r[asm.options.pure]

> The `pure` option may be specfied. The evaluation of the assembly block shall not produce any observable behaviour, consume input, or terminate execution, and the implementation may assume that the outputs of the assembly block depends only on the inputs and the contents of any allocation. If the program specifies the `pure` option, it shall specify either the `nomem` or `readonly` option.

Reviewer 2: ```suggestion
The `pure` option may be specified. The evaluation of the assembly block shall not produce any observable behaviour, consume input, or terminate execution, and the implementation may assume that the outputs of the assembly block depends only on the inputs and the contents of any allocation. If the program specifies the `pure` option, it shall specify either the `nomem` or `readonly` option.
```

> [!TARGET-SPECFIC]
> - These flags registers must be restored upon exiting the asm block if the `preserves_flags` option is set:
> - x86
>   - Status flags in `EFLAGS` (CF, PF, AF, ZF, SF, OF).
>   - Floating-point status word (all).
>   - Floating-point exception flags in `MXCSR` (PE, UE, OE, ZE, DE, IE).
> - ARM
>   - Condition flags in `CPSR` (N, Z, C, V)
>   - Saturation flag in `CPSR` (Q)
>   - Greater than or equal flags in `CPSR` (GE).
>   - Condition flags in `FPSCR` (N, Z, C, V)
>   - Saturation flag in `FPSCR` (QC)
>   - Floating-point exception flags in `FPSCR` (IDC, IXC, UFC, OFC, DZC, IOC).
> - AArch64
>   - Condition flags (`NZCV` register).
>   - Floating-point status (`FPSR` register).
> - RISC-V
>   - Floating-point exception flags in `fcsr` (`fflags`).
>   - Vector extension state (`vtype`, `vl`, `vcsr`).
> - LoongArch
>   - Floating-point condition flags in `$fcc[0-7]`.

```rust,ignore
# #[cfg(target_arch = "x86_64")] { unsafe{
core::arch::asm!("cmp eax, eax", in("eax") 5, options(preserve_flags));
# }}
```

r[asm.options.noreturn]

> The `noreturn` option may be specifed. An invocation of the [`core::arch::asm!`] macro that specifies the `noreturn` option expands to an expression of type `!`. The behaviour is undefined if an evaluation of the assembly block exits. The program shall not specify the `clobber_abi` specification, or an operand that is an `out`, `lateout`, `inout`, or `inlateout` operand. 

Reviewer 2: ```suggestion
The `noreturn` option may be specified. An invocation of the [`core::arch::asm!`] macro that specifies the `noreturn` option expands to an expression of type `!`. The behaviour is undefined if an evaluation of the assembly block exits. The program shall not specify the `clobber_abi` specification, or an operand that is an `out`, `lateout`, `inout`, or `inlateout` operand. 
```

> [!NOTE]
> This include state such as the current section of the assembler, the syntax mode, or the kind of assembly output being generated.
> Failing to obey this requirement can have significant impact on code generation, including code unrelated to the asm block. For example, an asm block that issues a `.data` directive without resetting to the appropriate section for the function can cause the following code in the function to be generated in the `.data` section, and for execution to fall off the asm block into improper memory.

r[asm.invocation.global-section]
The *expanded asm-string* of *global-asm-block* invocation acts as though a `.section` directive is issued before the *expanded asm-string*  which causes code to be generated in the default section on the target for executable code.

> [!NOTE]
> This section is typically named `.text`.

r[asm.invocation.prefix-instr]
An *expanded asm-string* shall not end with an instruction that is interpreted as a prefix on the architecture. No diagnostic is required.

> [!TARGET-SPECIFIC]

> > On x86 and x86-64, the `lock`, `repnz`, `rep`, `repz`, as well as GNU AS specific address-size, data-size, and explicit rex, vex, and evex prefixes.

Reviewer 2: This seems to be missing a verb. Is this saying something like:

```suggestion
> On x86 and x86-64, prefix instructions include the `lock`, `repnz`, `rep`, `repz`, as well as GNU AS specific address-size, data-size, and explicit rex, vex, and evex prefixes.
```

```rust
# #[cfg(target_arch = "x86_64")] { unsafe{
let x: i32 = 5;
let y: i32;
core::arch::asm!("mov {:e}, {:e}", out(reg) y, in(reg) x);
# }}
```

> [!NOTE]
> The supported modifiers are a subset of LLVM's (and GCC's) [asm template argument modifiers][llvm-argmod], but do not use the same letter codes.

> [!NOTE]
> - on ARM `e` / `f`: this prints the low or high doubleword register name of a NEON quad (128-bit) register.
> - on x86: our behavior for `reg` with no modifiers differs from what GCC does.
>   GCC will infer the modifier based on the operand value type, while we default to the full register size.
> - on x86 `xmm_reg`: the `x`, `t` and `g` LLVM modifiers are not yet implemented in LLVM (they are supported by GCC only), but this should be a simple change.

r[asm.template.diagnostic]
A lint diagnostic should be emitted if a modifier is omitted, or a modifier is used such that the modified expanded register is of an inappropriate width for the type used to initialize the operand

[llvm-argmod]: http://llvm.org/docs/LangRef.html#asm-template-argument-modifiers

## Behaviour of an asm block

r[asm.evaluation]

r[asm.evaluation.general]

> Each evaluation of an *asm block* shall perform an operation that correpsonds to the result of a valid sequence of operations on the Minirust Abstract Machine. The behaviour is undefined if the operations performed by the asm block do not validly correspond to a valid sequence of Minirust operations.

Reviewer 2: This seems like a new claim that we will be defining things in terms of Minirust. Is that something we have decided to do?

Author: So there are two separate "new claims" here - one is that the operational semantics will be defined in terms of Minirust in general (that had a whole meeting on it, where we took an hour and a half to say "Yes"... actually, no, we took 5 minutes to say "Yes" and then spent the other hour 25 minutes debating "to what extent"), and the other is that inline assembly must affect the Rust AM in a way that Rust Code could possibly have done.

I believe the latter has been the opinion of the opsem team since before it was the opsem team, and is more generally, makes the claim that compilers can optimize Rust in the face of unanalyzed code (which has been assumed as true since we started plugging Rust into llvm).

Author: In practice, this means that this code has UB, because it doesn't correspond back to a valid sequence of Minirust operations (that do not themselves have undefined behaviour):
```rust
let x: i32 = 0;
asm!("mov dword ptr [{}], 5", in(reg) &x);
println!("{x}"); // May be turned into `println!("0");`
```

The code writes to a `Frozen` tag, which is UB. But if inline assembly isn't constrained to simply doing things (to the AM) Rust code itself can do, then it could circumvent that rule.

Reviewer 2: > that had a whole meeting on it, where we took an hour and a half to say "Yes"... actually, no, we took 5 minutes to say "Yes" and then spent the other hour 25 minutes debating "to what extent")

I don't recall having a team decision on this point, and I don't see any discussions about that decision in the [meeting notes](https://hackmd.io/team/rust-spec-team?nav=overview). 

I'm not opposed to Minirust as an eventual basis to use for defining things, but I'm not yet convinced that we are in a position where we are ready to do that *today*.

I'd also say that adding things like this are pretty heavy decisions that are probably best not to bundle in one large update. This makes it harder to review and make progress when there are many new changes included in one PR. Is it possible to work in smaller batches, and to keep new additions separate from things that are more editorially oriented?

Author: I think we need to be able to say what an inline assembly block does, and I'm not sure there's any other model we can use. We know that it can do more than Rust code can.  For example, an asm block can yield a nondeterministic value, directly accessing the `pick` Minirust operation.

Reviewer 5: @ehuss is right here.  Maybe we do...

> need to be able to say what an inline assembly block does

...in more detail than what the Reference chapter says today.  But that doesn't need to all happen in this PR.  And trying to do that is making it difficult to review this and make progress here.

Reviewer 4: FWIW this paragraph is not specific to inline assembly, it generally applies to any kind of FFI call.

Author: Clarifying my statement at the T-spec meeting:
* I think we need *a* claim here, for the rest of the section and the following section ([asm.options]) to make sense.
* I don't think it necessarily needs to make a claim about Minirust.
* We know that the claim should allow *something* more than what Rust itself allows, but less than what the hardware allows.

> - `.cfi_adjust_cfa_offset`
> - `.cfi_def_cfa`
> - `.cfi_def_cfa_offset`
> - `.cfi_def_cfa_register`
> - `.cfi_endproc`
> - `.cfi_escape`
> - `.cfi_lsda`
> - `.cfi_offset`
> - `.cfi_personality`
> - `.cfi_register`
> - `.cfi_rel_offset`
> - `.cfi_remember_state`
> - `.cfi_restore`
> - `.cfi_restore_state`
> - `.cfi_return_column`
> - `.cfi_same_value`
> - `.cfi_sections`
> - `.cfi_signal_frame`
> - `.cfi_startproc`
> - `.cfi_undefined`
> - `.cfi_window_save`

r[asm.directives.seh]

> [!TARGET-SPECIFIC]
> The following directives are guaranteed to be supported on platforms that use Structured Exception Handling
> - `.seh_endproc`
> - `.seh_endprologue`
> - `.seh_proc`
> - `.seh_pushreg`
> - `.seh_savereg`
> - `.seh_setframe`
> - `.seh_stackalloc`

r[asm.directives.x86]

> [!TARGET-SPECIFIC]
> The following directives are guaranteed to be supported on x86 and x86-64 platforms
> - `.nops`
> - `.code16`
> - `.code32`
> - `.code64`
> Use of `.code16`, `.code32`, and `.code64` directives are only supported if the state is reset to the default before exiting the assembly block.
> 32-bit x86 uses `.code32` by default, and x86_64 uses `.code64` by default.

r[asm.directives.arm]

> > The following directives are guaranteed to be supported on 32-bit ARM platforms

Reviewer 3: Isn't this `[!TARGET-SPECIFIC]` too?
